File: modules/meeterm-terminal/android/src/main/java/dev/meeterm/terminal/AttachmentOperation.kt
package dev.meeterm.terminal

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

/**
 * Rust-owned attachment operation state (attachment-ffi contract).
 *
 * The core operation lifecycle is authoritative; this model only decodes the
 * fixed-size snapshot record and pins the adapter-side transition guards:
 * a second upload is refused while an op is live, snapshots from a stale
 * attachment id are dropped, and insert/delete capabilities follow the phase.
 */
internal enum class AttachmentOpPhase(val wireValue: Int) {
  PENDING(0),
  UPLOADING(1),
  UPLOADED(2),
  INSERTED(3),
  FAILED(4),
  CANCELLED(5),
  DELETED(6);

  companion object {
    fun fromWire(value: Int): AttachmentOpPhase? =
      values().firstOrNull { it.wireValue == value }
  }
}

internal data class AttachmentOperation(
  val attachmentId: Long,
  val phase: AttachmentOpPhase,
  val flags: Int,
  val bytesUploaded: Long,
  val sizeBytes: Long,
  val remotePath: String,
  val displayName: String,
  val errorCode: String,
  val errorMessage: String,
) {
  /** flags & 0x1: the input queue accepted the line; delivery unconfirmed. */
  val insertUnconfirmed: Boolean get() = flags and 0x1 != 0

  val wirePhase: String get() = when (phase) {
    AttachmentOpPhase.PENDING -> "pending"
    AttachmentOpPhase.UPLOADING -> "uploading"
    AttachmentOpPhase.UPLOADED -> "uploaded"
    AttachmentOpPhase.INSERTED -> "inserted"
    AttachmentOpPhase.FAILED -> "failed"
    AttachmentOpPhase.CANCELLED -> "cancelled"
    AttachmentOpPhase.DELETED -> "deleted"
  }

  /** Explicit Retry upload is offered only for pending/failed ops. */
  val canRetryUpload: Boolean
    get() = phase == AttachmentOpPhase.PENDING || phase == AttachmentOpPhase.FAILED

  /** Insert is allowed while uploaded; an inserted op accepts it idempotently. */
  val canInsert: Boolean
    get() = phase == AttachmentOpPhase.UPLOADED || phase == AttachmentOpPhase.INSERTED

  /** Cancel applies while transfer work may still be running. */
  val canCancel: Boolean
    get() = phase == AttachmentOpPhase.PENDING || phase == AttachmentOpPhase.UPLOADING

  /** M4 remote delete covers every phase that may own a completed file. */
  val canDeleteRemote: Boolean
    get() = phase == AttachmentOpPhase.UPLOADED ||
      phase == AttachmentOpPhase.INSERTED ||
      phase == AttachmentOpPhase.FAILED ||
      phase == AttachmentOpPhase.CANCELLED

  val isTerminal: Boolean
    get() = phase == AttachmentOpPhase.INSERTED ||
      phase == AttachmentOpPhase.DELETED
}

/**
 * Tracks the live core operation. The snapshot is authoritative, but adapter
 * guards keep a stale attachment id or a second upload from corrupting state.
 */
internal class AttachmentOpMachine {
  var operation: AttachmentOperation? = null
    private set

  /** A new upload may only replace a fully terminal (or absent) operation. */
  fun canBeginUpload(): Boolean {
    val op = operation ?: return true
    return op.phase == AttachmentOpPhase.CANCELLED ||
      op.phase == AttachmentOpPhase.DELETED ||
      op.phase == AttachmentOpPhase.FAILED
  }

  /** Record the attachment id accepted by the core; false means refused. */
  fun recordBegin(attachmentId: Long, sizeBytes: Long, displayName: String): Boolean {
    if (attachmentId <= 0 || !canBeginUpload()) return false
    operation = AttachmentOperation(
      attachmentId = attachmentId,
      phase = AttachmentOpPhase.UPLOADING,
      flags = 0,
      bytesUploaded = 0,
      sizeBytes = sizeBytes,
      remotePath = "",
      displayName = displayName,
      errorCode = "",
      errorMessage = "",
    )
    return true
  }

  /** Apply the authoritative core snapshot; stale ids are dropped. */
  fun applySnapshot(snapshot: AttachmentOperation): Boolean {
    val current = operation ?: return false
    if (current.attachmentId != snapshot.attachmentId) return false
    // A locally-cancelled op keeps its phase until the core confirms the
    // cancel (or finishes); late upload bytes never resurrect it.
    if (current.phase == AttachmentOpPhase.CANCELLED &&
      snapshot.phase == AttachmentOpPhase.UPLOADING
    ) {
      return true
    }
    operation = snapshot
    return true
  }

  /** Local cancel mark; the core drops delayed completions itself. */
  fun markCancelled() {
    operation = operation?.copy(phase = AttachmentOpPhase.CANCELLED)
  }

  fun clear() {
    operation = null
  }
}

/** Fixed-offset decoder for `meeterm_attachment_snapshot_t` (1000 bytes). */
internal object AttachmentOperationCodec {
  private const val REMOTE_PATH_LEN = 32
  private const val REMOTE_PATH = 34
  private const val DISPLAY_NAME_LEN = 546
  private const val DISPLAY_NAME = 548
  private const val ERROR_CODE_LEN = 676
  private const val ERROR_CODE = 678
  private const val ERROR_MESSAGE_LEN = 742
  private const val ERROR_MESSAGE = 744
  const val RECORD_SIZE = 1000

  fun decode(bytes: ByteArray): AttachmentOperation? {
    if (bytes.size < RECORD_SIZE) return null
    val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
    val phaseValue = buffer.getInt(0)
    val phase = AttachmentOpPhase.fromWire(phaseValue) ?: return null
    val flags = buffer.getInt(4)
    val attachmentId = buffer.getLong(8)
    val bytesUploaded = buffer.getLong(16)
    val sizeBytes = buffer.getLong(24)
    return AttachmentOperation(
      attachmentId = attachmentId,
      phase = phase,
      flags = flags,
      bytesUploaded = bytesUploaded,
      sizeBytes = sizeBytes,
      remotePath = field(buffer, bytes, REMOTE_PATH_LEN, REMOTE_PATH, 512),
      displayName = field(buffer, bytes, DISPLAY_NAME_LEN, DISPLAY_NAME, 128),
      errorCode = field(buffer, bytes, ERROR_CODE_LEN, ERROR_CODE, 64),
      errorMessage = field(buffer, bytes, ERROR_MESSAGE_LEN, ERROR_MESSAGE, 256),
    )
  }

  private fun field(
    buffer: ByteBuffer,
    bytes: ByteArray,
    lengthOffset: Int,
    dataOffset: Int,
    capacity: Int,
  ): String {
    val length = (buffer.getShort(lengthOffset).toInt() and 0xFFFF).coerceAtMost(capacity)
    val end = dataOffset + length
    if (end > bytes.size) return ""
    return String(bytes, dataOffset, length, StandardCharsets.UTF_8)
  }
}

File: modules/meeterm-terminal/ios/AttachmentOperation.swift
import Foundation

/**
 * Rust-owned attachment operation state (attachment-ffi contract).
 *
 * The core operation lifecycle is authoritative; this model only decodes the
 * fixed-size snapshot record and pins the adapter-side transition guards:
 * a second upload is refused while an op is live, snapshots from a stale
 * attachment id are dropped, and insert/delete capabilities follow the phase.
 */
enum AttachmentOpPhase: Int {
  case pending = 0
  case uploading = 1
  case uploaded = 2
  case inserted = 3
  case failed = 4
  case cancelled = 5

  var wireName: String {
    switch self {
    case .pending: return "pending"
    case .uploading: return "uploading"
    case .uploaded: return "uploaded"
    case .inserted: return "inserted"
    case .failed: return "failed"
    case .cancelled: return "cancelled"
    }
  }
}

struct AttachmentOperation: Equatable {
  let attachmentId: UInt64
  let phase: AttachmentOpPhase
  let flags: UInt32
  let bytesUploaded: UInt64
  let sizeBytes: UInt64
  let remotePath: String
  let displayName: String
  let errorCode: String
  let errorMessage: String

  /// flags & 0x1: the input queue accepted the line; delivery unconfirmed.
  var insertUnconfirmed: Bool { flags & 0x1 != 0 }

  /// flags & 0x2: the meeterm-created remote file was explicitly deleted.
  var remoteRemoved: Bool { flags & 0x2 != 0 }

  /// flags & 0x4: a job (upload / verify+insert / remove) is in flight on
  /// this operation. The core clears the previous reason at job start and
  /// drops the flag when that attempt's outcome lands — UI busy display
  /// and polling key off this bit, never off a stale errorCode.
  var jobInFlight: Bool { flags & 0x4 != 0 }

  /// Explicit Retry upload: pending/failed ops, plus an uploaded op whose
  /// remote file was deleted — the core re-uploads that same operation.
  /// Refused while a job is in flight.
  var canRetryUpload: Bool {
    !jobInFlight && (phase == .pending || phase == .failed ||
      (phase == .uploaded && remoteRemoved))
  }

  /// Insert is one verified job: intent check → fresh fence → remote
  /// lstat → single-line paste. It is allowed on an uploaded op only —
  /// an inserted op is never re-inserted — and never while another job
  /// runs.
  var canInsert: Bool {
    !remoteRemoved && !jobInFlight && phase == .uploaded
  }

  /// Cancel applies while transfer work may still be running.
  var canCancel: Bool { phase == .pending || phase == .uploading }

  /// Remote delete covers every phase that may still own a file; refused
  /// while a job is in flight (one in-flight job per operation).
  var canDeleteRemote: Bool {
    !remoteRemoved && !jobInFlight &&
      (phase == .uploaded || phase == .inserted || phase == .failed || phase == .cancelled)
  }
}

/**
 * Tracks the live core operation. The snapshot is authoritative, but adapter
 * guards keep a stale attachment id or a second upload from corrupting state.
 */
final class AttachmentOpMachine {
  private(set) var operation: AttachmentOperation?

  /// A new upload may only replace a fully terminal (or absent) operation.
  func canBeginUpload() -> Bool {
    guard let op = operation else { return true }
    return op.phase == .cancelled || op.phase == .failed
  }

  /// Record the attachment id accepted by the core; false means refused.
  @discardableResult
  func recordBegin(attachmentId: UInt64, sizeBytes: UInt64, displayName: String) -> Bool {
    guard attachmentId > 0, canBeginUpload() else { return false }
    operation = AttachmentOperation(
      attachmentId: attachmentId,
      phase: .uploading,
      flags: 0,
      bytesUploaded: 0,
      sizeBytes: sizeBytes,
      remotePath: "",
      displayName: displayName,
      errorCode: "",
      errorMessage: ""
    )
    return true
  }

  /// Apply the authoritative core snapshot; stale ids are dropped.
  @discardableResult
  func applySnapshot(_ snapshot: AttachmentOperation) -> Bool {
    guard let current = operation, current.attachmentId == snapshot.attachmentId else {
      return false
    }
    // A locally-cancelled op keeps its phase until the core confirms the
    // cancel (or finishes); late upload bytes never resurrect it.
    if current.phase == .cancelled, snapshot.phase == .uploading {
      return true
    }
    operation = snapshot
    return true
  }

  /// Local cancel mark; the core drops delayed completions itself.
  func markCancelled() {
    guard var op = operation else { return }
    op = AttachmentOperation(
      attachmentId: op.attachmentId,
      phase: .cancelled,
      flags: op.flags,
      bytesUploaded: op.bytesUploaded,
      sizeBytes: op.sizeBytes,
      remotePath: op.remotePath,
      displayName: op.displayName,
      errorCode: op.errorCode,
      errorMessage: op.errorMessage
    )
    operation = op
  }

  func clear() {
    operation = nil
  }
}

/// Fixed-offset decoder for `meeterm_attachment_snapshot_t` (1000 bytes).
enum AttachmentOperationCodec {
  static let recordSize = 1000

  static func decode(_ bytes: Data) -> AttachmentOperation? {
    guard bytes.count >= recordSize else { return nil }
    guard let phaseValue = readUInt32(bytes, at: 0),
          let phase = AttachmentOpPhase(rawValue: Int(phaseValue)) else { return nil }
    let flags = readUInt32(bytes, at: 4) ?? 0
    let attachmentId = readUInt64(bytes, at: 8) ?? 0
    let bytesUploaded = readUInt64(bytes, at: 16) ?? 0
    let sizeBytes = readUInt64(bytes, at: 24) ?? 0
    return AttachmentOperation(
      attachmentId: attachmentId,
      phase: phase,
      flags: flags,
      bytesUploaded: bytesUploaded,
      sizeBytes: sizeBytes,
      remotePath: field(bytes, lengthOffset: 32, dataOffset: 34, capacity: 512),
      displayName: field(bytes, lengthOffset: 546, dataOffset: 548, capacity: 128),
      errorCode: field(bytes, lengthOffset: 676, dataOffset: 678, capacity: 64),
      errorMessage: field(bytes, lengthOffset: 742, dataOffset: 744, capacity: 256)
    )
  }

  private static func readUInt32(_ data: Data, at offset: Int) -> UInt32? {
    guard offset + 4 <= data.count else { return nil }
    return data.withUnsafeBytes { raw in
      raw.load(fromByteOffset: offset, as: UInt32.self)
    }
  }

  private static func readUInt64(_ data: Data, at offset: Int) -> UInt64? {
    guard offset + 8 <= data.count else { return nil }
    return data.withUnsafeBytes { raw in
      raw.load(fromByteOffset: offset, as: UInt64.self)
    }
  }

  private static func field(
    _ data: Data,
    lengthOffset: Int,
    dataOffset: Int,
    capacity: Int
  ) -> String {
    guard lengthOffset + 2 <= data.count else { return "" }
    let declared = data.withUnsafeBytes { raw -> Int in
      Int(raw.load(fromByteOffset: lengthOffset, as: UInt16.self))
    }
    let length = min(declared, capacity)
    guard dataOffset + length <= data.count else { return "" }
    return String(decoding: data[dataOffset ..< dataOffset + length], as: UTF8.self)
  }
}
